### DISCUSSION TITLE
Allow `typeof(this)`

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#7013] [Language Ideas] [answered]
Imagine a generic class:
```cs
class SomeLongAndComplicatedClassName<TArgument1, TArgument2, TArgument3>
{
  void Foo()
  {
    // the next line is unwieldy and error prone
    Type t = typeof(SomeLongAndComplicatedClassName<TArgument1, TArgument2, TArgument3>); 
    // on the other hand, this would be much better
    Type t2 = typeof(this); 
  }
}
```
Basically, `typeof(this)` is the compile time type of the containing type, and works in both instance and static contexts, in methods/props and field initializers. (Not to be mistaken with `this.GetType()`!)

**Comment by jnm2** [comment, marked as answer]
Subset of https://github.com/dotnet/csharplang/discussions/5261

**Reply by TahirAhmadov** [reply]
Thanks, I searched but couldn't find that discussion.
